Source/WebCore/dom/ContentVisibilityDocumentState.cpp

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (C) 2023 Igalia S.L. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "ContentVisibilityDocumentState.h"
+
+#include "Element.h"
+#include "IntersectionObserverCallback.h"
+#include "NodeRenderStyle.h"
+#include "RenderElement.h"
+#include "RenderStyle.h"
+
+namespace WebCore {
+
+class ContentVisibilityIntersectionObserverCallback final : public IntersectionObserverCallback {
+public:
+    static Ref<ContentVisibilityIntersectionObserverCallback> create(Document& document)
+    {
+        return adoptRef(*new ContentVisibilityIntersectionObserverCallback(document));
+    }
+
+private:
+    bool hasCallback() const final { return true; }
+
+    CallbackResult<void> handleEvent(IntersectionObserver&, const Vector<Ref<IntersectionObserverEntry>>& entries, IntersectionObserver&) final
+    {
+        ASSERT(!entries.isEmpty());
+
+        for (auto& entry : entries) {
+            if (auto* element = entry->target()) {
+                element->contentVisibilityViewportChange(entry->isIntersecting());
+                element->document().contentVisibilityDocumentState().updateOnScreenObservationTarget(*element, entry->isIntersecting());
+            }
+        }
+        return { };
+    }
+
+    ContentVisibilityIntersectionObserverCallback(Document& document)
+        : IntersectionObserverCallback(&document)
+    {
+    }
+};
+
+void ContentVisibilityDocumentState::observe(Element& element)
+{
+    auto& state = element.document().contentVisibilityDocumentState();
```

Comment:
Isn't this just referencing itself? Should we assert the element.document() here?

Reply:
This is done this way since observe/unobserve are static methods.